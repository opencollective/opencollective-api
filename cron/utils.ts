import config from 'config';

/**
 * Heroku scheduler only has daily or hourly cron jobs, we only want to run
 * this script once per week on Monday (1). If the day is not Monday on production
 * we won't execute the script
 */
export function onlyExecuteInProdOnMondays() {
  const today = new Date();
  if (config.env === 'production' && today.getDay() !== 1) {
    console.log('OC_ENV is production and day is not Monday, script aborted!');
    process.exit(0);
  }
}
