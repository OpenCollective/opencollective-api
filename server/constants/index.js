import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, roles, math, channels, transactions };
