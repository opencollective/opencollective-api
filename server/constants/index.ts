import activities from './activities';
import channels from './channels';
import expenseStatus from './expense_status';
import expenseTypes from './expense_type';
import frequencies from './frequencies';
import math from './math';
import roles from './roles';
import * as transactions from './transactions';

export { activities, expenseStatus, expenseTypes, roles, math, channels, transactions, frequencies };
