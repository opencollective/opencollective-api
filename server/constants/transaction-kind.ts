export enum TransactionKind {
  /** Transactions coming from the "Add Funds" */
  ADDED_FUNDS = 'ADDED_FUNDS',
  /** Transactions from transferring the remaining balance from a project/event/collective **/
  BALANCE_TRANSFER = 'BALANCE_TRANSFER',
  /** Transactions coming from the "Contribution Flow" */
  CONTRIBUTION = 'CONTRIBUTION',
  /** Transactions coming from the "Expense Flow" */
  EXPENSE = 'EXPENSE',
  /** The host fee going to the Host */
  HOST_FEE = 'HOST_FEE',
  /** Part of the Host fee going from the Host to the Platform */
  HOST_FEE_SHARE = 'HOST_FEE_SHARE',
  /** Part of the Host fee going from the Host to the Platform */
  HOST_FEE_SHARE_DEBT = 'HOST_FEE_SHARE_DEBT',
  /** Amount given by Fiscal Hosts to cover payment processor fee on refunds */
  PAYMENT_PROCESSOR_COVER = 'PAYMENT_PROCESSOR_COVER',
  /** Reserved keyword in case we want to use in the future */
  PAYMENT_PROCESSOR_FEE = 'PAYMENT_PROCESSOR_FEE',
  /** Reserved keyword in case we want to use in the future */
  PLATFORM_FEE = 'PLATFORM_FEE',
  /** Financial contribution to Open Collective added on top of another contribution */
  PLATFORM_TIP = 'PLATFORM_TIP',
  /** Financial contribution to Open Collective added on top of another contribution */
  PLATFORM_TIP_DEBT = 'PLATFORM_TIP_DEBT',
  /** For prepaid budgets */
  PREPAID_PAYMENT_METHOD = 'PREPAID_PAYMENT_METHOD',
}
