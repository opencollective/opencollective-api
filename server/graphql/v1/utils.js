import debugLib from 'debug';
import { graphql } from 'graphql';

import { loaders } from '../loaders';

import schema from './schema';

const debug = debugLib('graphql');

export const makeRequest = (remoteUser, query) => {
  return {
    remoteUser,
    body: { query },
    loaders: loaders({ remoteUser }),
  };
};

export const graphqlQuery = async (query, variables, remoteUser) => {
  const prepare = () => {
    if (remoteUser) {
      remoteUser.rolesByCollectiveId = null; // force refetching the roles
      return remoteUser.populateRoles();
    } else {
      return Promise.resolve();
    }
  };

  if (process.env.DEBUG && process.env.DEBUG.match(/graphql/)) {
    debug('query', query);
    debug('variables', variables);
    debug('context', remoteUser);
  }

  return prepare().then(() =>
    graphql({
      schema,
      source: query,
      rootValue: null,
      contextValue: makeRequest(remoteUser, query),
      variableValues: variables,
    }),
  );
};
