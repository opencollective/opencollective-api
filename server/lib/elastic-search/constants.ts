export enum ElasticSearchIndexName {
  COLLECTIVES = 'collectives',
  COMMENTS = 'comments',
  EXPENSES = 'expenses',
  UPDATES = 'updates',
  TRANSACTIONS = 'transactions',
  ORDERS = 'orders',
  TIERS = 'tiers',
  HOST_APPLICATIONS = 'host-applications',
}
