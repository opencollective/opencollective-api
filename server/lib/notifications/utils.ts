import sanitizeHtml from 'sanitize-html';

import { sanitizerOptions as updateSanitizerOptions } from '../../models/Update';

const constructPreviewImageURL = (service: string, id: string) => {
  if (service === 'youtube' && id.match('[a-zA-Z0-9_-]{11}')) {
    return `https://img.youtube.com/vi/${id}/0.jpg`;
  } else if (service === 'anchorFm') {
    return `https://opencollective.com/static/images/anchor-fm-logo.png`;
  } else {
    return null;
  }
};

const parseServiceLink = (videoLink: string) => {
  const regexps = {
    youtube: new RegExp(
      '(?:https?://)?(?:www\\.)?youtu(?:\\.be/|be(-nocookie)?\\.com/\\S*(?:watch|embed)(?:(?:(?=/[^&\\s?]+(?!\\S))/)|(?:\\S*v=|v/)))([^&\\s?]+)',
      'i',
    ),
    anchorFm: /^(http|https)?:\/\/(www\.)?anchor\.fm\/([^/]+)(\/embed)?(\/episodes\/)?([^/]+)?\/?$/,
  };
  for (const service in regexps) {
    videoLink = videoLink.replace('/?showinfo=0', '');
    const matches = regexps[service].exec(videoLink);
    if (matches) {
      if (service === 'anchorFm') {
        const podcastName = matches[3];
        const episodeId = matches[6];
        const podcastUrl = `${podcastName}/embed`;
        return { service, id: episodeId ? `${podcastUrl}/episodes/${episodeId}` : podcastUrl };
      } else {
        return { service, id: matches[matches.length - 1] };
      }
    }
  }
  return {};
};

export const replaceVideosByImagePreviews = (html: string) => {
  const sanitizerOptions = {
    ...updateSanitizerOptions,
    transformTags: {
      ...updateSanitizerOptions.transformTags,
      iframe: (tagName, attribs) => {
        if (!attribs.src) {
          return '';
        }
        const { service, id } = parseServiceLink(attribs.src);
        const imgSrc = constructPreviewImageURL(service, id);
        if (imgSrc) {
          return {
            tagName: 'img',
            attribs: {
              src: imgSrc,
              alt: `${service} content`,
            },
          };
        } else {
          return '';
        }
      },
    },
  };
  return sanitizeHtml(html, sanitizerOptions);
};
