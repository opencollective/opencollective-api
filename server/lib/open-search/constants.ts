import { CollectiveType } from '../../constants/collectives';

export enum OpenSearchIndexName {
  COLLECTIVES = 'collectives',
  COMMENTS = 'comments',
  EXPENSES = 'expenses',
  UPDATES = 'updates',
  TRANSACTIONS = 'transactions',
  ORDERS = 'orders',
  TIERS = 'tiers',
  HOST_APPLICATIONS = 'host-applications',
}

export interface OpenSearchIndexParams extends Record<OpenSearchIndexName, Record<string, unknown>> {
  [OpenSearchIndexName.COLLECTIVES]: {
    type?: CollectiveType;
    isHost?: boolean;
    tags?: string[];
  };
}
