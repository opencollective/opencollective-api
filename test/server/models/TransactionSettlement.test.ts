import { expect } from 'chai';
import { times } from 'lodash';

import PlatformConstants from '../../../server/constants/platform';
import { TransactionKind } from '../../../server/constants/transaction-kind';
import models, { Op } from '../../../server/models';
import TransactionSettlement, { TransactionSettlementStatus } from '../../../server/models/TransactionSettlement';
import {
  fakeCollective,
  fakeExpense,
  fakeHost,
  fakeOrganization,
  fakeTransaction,
  fakeUser,
  fakeUUID,
} from '../../test-helpers/fake-data';
import * as utils from '../../utils';

// ---- Helpers ----

const fakeContribution = (fromCollective, collective, transactionGroup) => {
  return fakeTransaction(
    {
      TransactionGroup: fakeUUID(transactionGroup),
      kind: TransactionKind.CONTRIBUTION,
      CollectiveId: collective.id,
      HostCollectiveId: collective.HostCollectiveId,
      FromCollectiveId: fromCollective.id,
      amount: 1000,
      type: 'CREDIT',
      description: 'A simple contribution **without** platform tip',
    },
    {
      createDoubleEntry: true,
    },
  );
};

// Insert contributions **with platform tips**, using the new format described in https://github.com/opencollective/opencollective/issues/4124.
// As writing this, the rest of the code is not yet using this new format.
const fakeContributionWithPlatformTipNewFormat = async (fromCollective, collective, transactionGroup) => {
  // Base transaction
  const transaction = await fakeTransaction(
    {
      TransactionGroup: fakeUUID(transactionGroup),
      kind: TransactionKind.CONTRIBUTION,
      FromCollectiveId: fromCollective.id,
      CollectiveId: collective.id,
      HostCollectiveId: collective.isHostAccount ? collective.id : collective.HostCollectiveId,
      amount: 1000,
      description: 'A contribution **with** platform tip',
    },
    {
      createDoubleEntry: true,
    },
  );

  // DEBIT Tip transaction (contributor -> OC)
  await fakeTransaction({
    TransactionGroup: transaction.TransactionGroup,
    kind: TransactionKind.PLATFORM_TIP,
    type: 'DEBIT',
    amount: -200,
    CollectiveId: transaction.FromCollectiveId,
    FromCollectiveId: PlatformConstants.PlatformCollectiveId,
    HostCollectiveId: null,
    description: 'Tip transaction from the contributor to Open Collective',
  });

  // CREDIT Tip transaction (contributor -> OC)
  await fakeTransaction({
    TransactionGroup: transaction.TransactionGroup,
    kind: TransactionKind.PLATFORM_TIP,
    type: 'CREDIT',
    amount: 200,
    FromCollectiveId: transaction.FromCollectiveId,
    CollectiveId: PlatformConstants.PlatformCollectiveId,
    HostCollectiveId: PlatformConstants.PlatformCollectiveId,
    description: 'Tip transaction from the contributor to Open Collective',
  });

  // [Debt] DEBIT Tip debt transaction (host -> OC)
  await fakeTransaction({
    kind: TransactionKind.PLATFORM_TIP_DEBT,
    TransactionGroup: transaction.TransactionGroup,
    type: 'DEBIT',
    amount: -200,
    FromCollectiveId: transaction.HostCollectiveId,
    CollectiveId: PlatformConstants.PlatformCollectiveId,
    HostCollectiveId: transaction.HostCollectiveId,
    isDebt: true,
    description: 'Tip transaction from the HOST to Open Collective',
  });

  // [Debt] CREDIT Tip debt transaction (host -> OC)
  const tipDebtCreditTransaction = await fakeTransaction({
    kind: TransactionKind.PLATFORM_TIP_DEBT,
    type: 'CREDIT',
    TransactionGroup: transaction.TransactionGroup,
    amount: 200,
    FromCollectiveId: PlatformConstants.PlatformCollectiveId,
    CollectiveId: transaction.HostCollectiveId,
    HostCollectiveId: PlatformConstants.PlatformCollectiveId,
    isDebt: true,
    description: 'Tip transaction from the HOST to Open Collective',
  });

  // [Debt] Transaction Settlement
  await models.TransactionSettlement.createForTransaction(tipDebtCreditTransaction);
};

const SNAPSHOT_COLUMNS = [
  'TransactionGroup',
  'type',
  'kind',
  'amount',
  'CollectiveId',
  'FromCollectiveId',
  'HostCollectiveId',
  'isDebt',
  'isRefund',
  'settlementStatus',
  'description',
];

// ---- Test ----

describe('server/models/TransactionSettlement', () => {
  let host, collective;

  before(async () => {
    await utils.resetTestDB();

    await fakeOrganization({ id: PlatformConstants.PlatformCollectiveId, name: 'Open Collective' });
    host = await fakeHost({ name: 'Open Source' });
    collective = await fakeCollective({ HostCollectiveId: host.id, name: 'ESLint' });

    // Insert unrelated transactions (to make sure they won't appear in the results)
    await Promise.all(
      times(5, () =>
        fakeTransaction({ description: 'Random contribution to another collective' }, { createDoubleEntry: true }),
      ),
    );

    // Insert contributions to host and hosted collectives
    const fromUser = await fakeUser(undefined, { name: 'Benjamin' });
    await fakeContribution(fromUser.collective, host, '00000001');
    await fakeContribution(fromUser.collective, collective, '00000002');

    // Insert contributions with platform tips to host and hosted collectives
    await fakeContributionWithPlatformTipNewFormat(fromUser.collective, host, '00000003');
    await fakeContributionWithPlatformTipNewFormat(fromUser.collective, collective, '00000004');
  });

  it('0. Properly initializes test data', async () => {
    // Snapshot initial state for simpler reviews & debug. Ignore unrelated transactions
    const watchedCollectiveIds = [collective.id, host.id, PlatformConstants.PlatformCollectiveId];
    const transactions = await models.Transaction.findAll({
      include: [{ association: 'host' }, { association: 'collective' }, { association: 'fromCollective' }],
      order: [['id', 'ASC']],
      where: {
        [Op.or]: [
          { CollectiveId: watchedCollectiveIds },
          { FromCollectiveId: watchedCollectiveIds },
          { HostCollectiveId: watchedCollectiveIds },
        ],
      },
    });

    await TransactionSettlement.attachStatusesToTransactions(transactions);
    utils.snapshotTransactions(transactions, { columns: SNAPSHOT_COLUMNS });
  });

  it('1. getHostDebts returns all debts for host', async () => {
    const debts = await TransactionSettlement.getHostDebts(host.id);
    await utils.preloadAssociationsForTransactions(debts, SNAPSHOT_COLUMNS);
    utils.snapshotTransactions(debts, { columns: SNAPSHOT_COLUMNS, loadAssociations: true });
  });

  it('2. updateTransactionsSettlementStatus updates statuses selectively', async () => {
    const debts = await TransactionSettlement.getHostDebts(host.id);
    const debtToUpdate = debts[0]; // Only settle the first debt
    const newStatus = TransactionSettlementStatus.SETTLED;
    const settlementExpense = await fakeExpense();
    await TransactionSettlement.updateTransactionsSettlementStatus([debtToUpdate], newStatus, settlementExpense.id);

    const updatedDebts = await TransactionSettlement.getHostDebts(host.id);
    await utils.preloadAssociationsForTransactions(updatedDebts, SNAPSHOT_COLUMNS);
    utils.snapshotTransactions(updatedDebts, { columns: SNAPSHOT_COLUMNS, loadAssociations: true });
  });

  it('3. getHostDebts can then filter by settlement status', async () => {
    const debts = await TransactionSettlement.getHostDebts(host.id, TransactionSettlementStatus.OWED);
    await utils.preloadAssociationsForTransactions(debts, SNAPSHOT_COLUMNS);
    utils.snapshotTransactions(debts, { columns: SNAPSHOT_COLUMNS, loadAssociations: true });
  });

  it('4. getAccountsWithOwedSettlements returns all accounts with OWED debts', async () => {
    const collectives = await TransactionSettlement.getAccountsWithOwedSettlements();
    expect(collectives.length).to.eq(1);
    expect(collectives[0].id).to.eq(host.id);
  });
});
