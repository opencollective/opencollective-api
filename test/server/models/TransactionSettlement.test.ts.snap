// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server/models/TransactionSettlement 0. Properly initializes test data 1`] = `
"
| Group     | type   | kind              | amount | To              | From            | Host            | isDebt | isRefund | Settlement | description                                             |
| --------- | ------ | ----------------- | ------ | --------------- | --------------- | --------------- | ------ | -------- | ---------- | ------------------------------------------------------- |
| #00000001 | DEBIT  | CONTRIBUTION      | -1000  | Benjamin        | Open Source     | NULL            | false  | false    |            | A simple contribution **without** platform tip          |
| #00000001 | CREDIT | CONTRIBUTION      | 1000   | Open Source     | Benjamin        | NULL            | false  | false    |            | A simple contribution **without** platform tip          |
| #00000002 | DEBIT  | CONTRIBUTION      | -1000  | Benjamin        | ESLint          | NULL            | false  | false    |            | A simple contribution **without** platform tip          |
| #00000002 | CREDIT | CONTRIBUTION      | 1000   | ESLint          | Benjamin        | Open Source     | false  | false    |            | A simple contribution **without** platform tip          |
| #00000003 | DEBIT  | CONTRIBUTION      | -1000  | Benjamin        | Open Source     | NULL            | false  | false    |            | A contribution **with** platform tip                    |
| #00000003 | CREDIT | CONTRIBUTION      | 1000   | Open Source     | Benjamin        | Open Source     | false  | false    |            | A contribution **with** platform tip                    |
| #00000003 | DEBIT  | PLATFORM_TIP      | -200   | Benjamin        | Open Collective | NULL            | false  | false    |            | Tip transaction from the contributor to Open Collective |
| #00000003 | CREDIT | PLATFORM_TIP      | 200    | Open Collective | Benjamin        | Open Collective | false  | false    |            | Tip transaction from the contributor to Open Collective |
| #00000003 | DEBIT  | PLATFORM_TIP_DEBT | -200   | Open Collective | Open Source     | Open Source     | true   | false    | OWED       | Tip transaction from the HOST to Open Collective        |
| #00000003 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source     | Open Collective | Open Collective | true   | false    | OWED       | Tip transaction from the HOST to Open Collective        |
| #00000004 | DEBIT  | CONTRIBUTION      | -1000  | Benjamin        | ESLint          | NULL            | false  | false    |            | A contribution **with** platform tip                    |
| #00000004 | CREDIT | CONTRIBUTION      | 1000   | ESLint          | Benjamin        | Open Source     | false  | false    |            | A contribution **with** platform tip                    |
| #00000004 | DEBIT  | PLATFORM_TIP      | -200   | Benjamin        | Open Collective | NULL            | false  | false    |            | Tip transaction from the contributor to Open Collective |
| #00000004 | CREDIT | PLATFORM_TIP      | 200    | Open Collective | Benjamin        | Open Collective | false  | false    |            | Tip transaction from the contributor to Open Collective |
| #00000004 | DEBIT  | PLATFORM_TIP_DEBT | -200   | Open Collective | Open Source     | Open Source     | true   | false    | OWED       | Tip transaction from the HOST to Open Collective        |
| #00000004 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source     | Open Collective | Open Collective | true   | false    | OWED       | Tip transaction from the HOST to Open Collective        |"
`;

exports[`server/models/TransactionSettlement 1. getHostDebts returns all debts for host 1`] = `
"
| Group     | type   | kind              | amount | To          | From            | Host            | isDebt | isRefund | Settlement | description                                      |
| --------- | ------ | ----------------- | ------ | ----------- | --------------- | --------------- | ------ | -------- | ---------- | ------------------------------------------------ |
| #00000003 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source | Open Collective | Open Collective | true   | false    | OWED       | Tip transaction from the HOST to Open Collective |
| #00000004 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source | Open Collective | Open Collective | true   | false    | OWED       | Tip transaction from the HOST to Open Collective |"
`;

exports[`server/models/TransactionSettlement 2. updateTransactionsSettlementStatus updates statuses selectively 1`] = `
"
| Group     | type   | kind              | amount | To          | From            | Host            | isDebt | isRefund | Settlement | description                                      |
| --------- | ------ | ----------------- | ------ | ----------- | --------------- | --------------- | ------ | -------- | ---------- | ------------------------------------------------ |
| #00000003 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source | Open Collective | Open Collective | true   | false    | SETTLED    | Tip transaction from the HOST to Open Collective |
| #00000004 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source | Open Collective | Open Collective | true   | false    | OWED       | Tip transaction from the HOST to Open Collective |"
`;

exports[`server/models/TransactionSettlement 3. getHostDebts can then filter by settlement status 1`] = `
"
| Group     | type   | kind              | amount | To          | From            | Host            | isDebt | isRefund | Settlement | description                                      |
| --------- | ------ | ----------------- | ------ | ----------- | --------------- | --------------- | ------ | -------- | ---------- | ------------------------------------------------ |
| #00000004 | CREDIT | PLATFORM_TIP_DEBT | 200    | Open Source | Open Collective | Open Collective | true   | false    | OWED       | Tip transaction from the HOST to Open Collective |"
`;
